### PR TITLE
Allow more fine-grained control over tokenizer input

### PR DESCRIPTION
**Pull request by danieldk**
## Description

Before this change, tokenizers took `List[str]` is their input. However, this is not ideal in some cases. For instance in many LLMs we have to provide a model with special pieces to signal e.g. the start/end of a prompt. To serve these use cases, this change allows more fine-grained control over the input.

Besides supporting `List[str]`, the tokenizers now also support `List[InputChunks]`. `InputChunks` is a list-derived data structure that can contain one of two types of chunks:

* `TextChunk`: this is a regular piece of text that should be tokenized.
* `SpecialPieceChunk`: this type of chunk stores a single piece that should be looked up in directly in the vocabulary. Additionally, it can have strings to be appended/prepended to the preceding/succeeding text chunks.

The tokenizer normalizes these chunks by merging consecutive text chunks and adding the strings specified in special piece chunks. After this merge step, it processes the chunks.

Since we can have special pieces in the input with this change, begin/end of sequence markers can now be added safely in the pre-encoders.

In order to test the use of special piece chunks in the input, the GPT-NeoX tokenizer is updated to include special pieces in the vocabulary. We may eventually want to do this for other types of tokenizers as well.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by danieldk** [comment]
> Looks like I am relying on features that are too now, I'll check if there is an easy fix.

`typing` saves the day.

File: curated_transformers/tokenization/chunks.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        new_chunks = MergedInputChunks()
+        for chunk in self:
+            last_is_text = new_chunks and isinstance(new_chunks[-1], TextChunk)
+            if isinstance(chunk, TextChunk):
+                if last_is_text:
+                    new_chunks[-1].text += chunk.text  # type: ignore[union-attr]
+                else:
+                    new_chunks.append(dataclasses.replace(chunk))
+            else:
```
Comment (shadeMe) [inline curated_transformers/tokenization/chunks.py:83]: It'd be a good idea to do a `all((isinstance(x, (SpecialPieceChunk, TextChunk) for x in self))` at the beginning to check if all the elements are of the expected type(s).

File: curated_transformers/tokenization/roberta_tokenizer.py [hunk 2]
```diff
@@ -4,9 +4,10 @@
 from pathlib import Path
 
 from .bbpe_tokenizer import ByteBPETokenizer
+from .chunks import InputChunks, SpecialPieceChunk, TextChunk
```
Comment (shadeMe) [inline curated_transformers/tokenization/roberta_tokenizer.py:7]:
```suggestion
```

File: curated_transformers/tokenization/roberta_tokenizer.py [hunk 3]
```diff
@@ -4,9 +4,10 @@
...
 from .hf_hub import FromPretrainedHFTokenizer
-from .tokenizer import PiecesWithIds, PostEncoder, PreDecoder
-from .util import remove_pieces_from_sequence, add_bos_eos_to_encoding
+from .tokenizer import AddBosEosPreEncoder, PreEncoder, PreDecoder
```
Comment (shadeMe) [inline curated_transformers/tokenization/roberta_tokenizer.py:9]:
```suggestion
from .tokenizer import AddBosEosPreEncoder, PreDecoder
```

File: curated_transformers/tokenization/gpt_neox_tokenizer.py [hunk 4]
```diff
@@ -39,7 +39,11 @@ def convert_hf_tokenizer(cls: Type[Self], tokenizer: Any) -> Self:
         deserialized = json.loads(serialized)
         vocab_merges = deserialized["model"]
         merges = [tuple(merge.split(" ")) for merge in vocab_merges["merges"]]
-        processor = ByteBPEProcessor(vocab_merges["vocab"], merges)
+        vocab = vocab_merges["vocab"]
```
Comment (shadeMe) [inline curated_transformers/tokenization/gpt_neox_tokenizer.py:42]: Don't we need to add this logic to other `ByteBPETokenizer` subclasses as well?

File: curated_transformers/tokenization/bbpe_tokenizer.py [hunk 5]
```diff
@@ -22,22 +27,30 @@ def __init__(
...
-                text_pieces.extend(token_pieces)
-
-            ids.append(text_ids)
-            pieces.append(text_pieces)
+        for seq in input:
+            seq_ids = []
+            seq_pieces = []
+
+            for chunk in seq:
+                if isinstance(chunk, MergedSpecialPieceChunk):
```
Comment (shadeMe) [inline curated_transformers/tokenization/bbpe_tokenizer.py:39]: Do we want to add validate the iterable's elements' type here (and in other tokenizers) as well?

File: docs/source/tokenizers.rst [hunk 6]
```diff
@@ -1,6 +1,30 @@
 Tokenization
 ============
 
+Tokenizer inputs
+----------------
+
+Each tokenizer accepts a ``List[str]`` or a ``List[InputChunks]``. In
+most cases passing a list of strings suffices. However passing
```
Comment (shadeMe) [inline docs/source/tokenizers.rst:8]:
```suggestion
most cases, passing a list of strings should suffice. However, passing
```

File: docs/source/tokenizers.rst [hunk 7]
```diff
@@ -1,6 +1,30 @@
...
+:class:`.InputChunks` can be useful when adding special pieces to the input is
+required.
```
Comment (shadeMe) [inline docs/source/tokenizers.rst:10]:
```suggestion
:class:`.InputChunks` can be useful when special pieces need to be added
to the input.
```

File: docs/source/tokenizers.rst [hunk 8]
```diff
@@ -1,6 +1,30 @@
...
+
+When the tokenizer is called with a list of strings, each string is
+automatically converted to a :class:`.TextChunk`, which represents a text chunk
+that should be tokenized. The other type of supported chunk is the
+:class:`.SpecialPieceChunk`. The piece stored by this type of chunk is not
+tokenized, but looked up directly.
```
Comment (shadeMe) [inline docs/source/tokenizers.rst:16]:
```suggestion
:class:`.SpecialPieceChunk`. The piece stored by this type of chunk is not
tokenized but looked up directly.
```

File: curated_transformers/tokenization/chunks.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        in special piece chunks.
+        """
+        new_chunks = MergedInputChunks()
+        for chunk in self:
+            last_is_text = new_chunks and isinstance(new_chunks[-1], TextChunk)
+            if isinstance(chunk, TextChunk):
+                if last_is_text:
+                    new_chunks[-1].text += chunk.text  # type: ignore[union-attr]
+                else:
+                    new_chunks.append(dataclasses.replace(chunk))
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/tokenization/chunks.py:82]: Yeah, that's apparently the standard way to copy with `dataclasses`, but it's quite handy, you can optionally add replacements as kwargs.

File: curated_transformers/tokenization/gpt_neox_tokenizer.py (hunk 4 above)
Comment (danieldk) [reply curated_transformers/tokenization/gpt_neox_tokenizer.py:42]: Definitely. I want to do this in a follow up PR, where we also switch to `tokenizer.json` for the non-sentencepiece tokenizers.

File: curated_transformers/tokenization/bbpe_tokenizer.py (hunk 5 above)
Comment (shadeMe) [reply curated_transformers/tokenization/bbpe_tokenizer.py:39]: Ah, right. That's true.

**Review by shadeMe — APPROVED** [review]

**Pull request merged**